### PR TITLE
Remove hardcoded specs

### PR DESCRIPTION
**Pull request by ZeeshanM96**
### Description:
In the ServiceArgs class we have several hardcoded sections for downloading and setting chain-spec files. These have been replaced by a config instead and can be removed.

Task card:  [1885](https://nextcloud.dwellir.com/apps/deck/#/board/29/card/1885)

### Test:
- Test the config:
Screenshot from 2024-05-15 18-07-19

- Config applied:
Screenshot from 2024-05-15 18-08-00

### CODE REVIEW

File: src/service_args.py [hunk 1]
```diff
@@ -161,132 +123,4 @@ def __customize_service_args(self):
...
-        chain_json_path = f'{c.CHAIN_SPEC_PATH}/{self.chain_name}.json'
-        utils.download_chain_spec(chain_json_url, f'{self.chain_name}.json')
-        self.__set_chain_name(chain_json_path, 0)
-
-    def __liberland(self):
-        chain_json_url = 'https://raw.githubusercontent.com/liberland/liberland_substrate/main/substrate/specs/mainnet.json'
-        chain_json_path = f'{c.CHAIN_SPEC_PATH}/{self.chain_name}.json'
-        utils.download_chain_spec(chain_json_url, f'{self.chain_name}.json')
-        self.__set_chain_name(chain_json_path, 0)
+            self.__add_firstchain_args(['--wasm-runtime-overrides', c.WASM_PATH])
```
Comment (jakobilobi) [inline src/service_args.py:126]: Why remove empty line at end of file?

**Review by jakobilobi — COMMENTED** [review]
We'll need to be careful when this later gets merged into `main` and we start refreshing charms.

Is it a part of the task, or does a separate task exist, to populate e.g. our Collectives with all the chainspec specifics we remove from here? Do you know? @jonathanudd

File: src/service_args.py (hunk 1 above)
Comment (ZeeshanM96) [reply src/service_args.py]: I was adding missing `)` in the end and removed the empty line as well. Anyhow it's fixed now!

**Comment by ZeeshanM96** [comment]
> We'll need to be careful when this later gets merged into `main` and we start refreshing charms.
> 
> Is it a part of the task, or does a separate task exist, to populate e.g. our Collectives with all the chainspec specifics we remove from here? Do you know? @jonathanudd

[Card](https://nextcloud.dwellir.com/apps/deck/#/board/29/card/1885) says downloading and setting chain-spec files have been replaced by a config instead and can be removed.

File: src/service_args.py [hunk 2]
```diff
@@ -162,131 +124,4 @@ def __customize_service_args(self):
...
-        self.__set_chain_name(chain_json_path, 0)
-
-    def __equilibrium(self):
-        self.__set_chain_name(Path(c.HOME_PATH, 'chainspec.json'), 0)
-
-    def __aleph_zero(self):
-        if self.chain_name.endswith('testnet'):
-            self.__set_chain_name('testnet', 0)
-        elif self.chain_name.endswith('mainnet'):
-            self.__set_chain_name('mainnet', 0)
```
[the remark this answers is not in the document]
Comment (ZeeshanM96) [reply src/service_args.py]: Thanks for the feedback and understood @jonathanudd , I have updated the code as you suggested. Please have a look!

**Comment by jakobilobi** [comment]
> [Card](https://nextcloud.dwellir.com/apps/deck/#/board/29/card/1885) says downloading and setting chain-spec files have been replaced by a config instead and can be removed.

It doesn't mean they have actually been set anywhere in production. My guess is that many charms still rely on these hard-coded values, and would perhaps break if this is pushed without some config updates preceding it.

> I guess it doesn't hurt to add the chain spec urls to the collective pages. But they can also be found in the config of the deployments and in the [bundle files](https://github.com/dwellir-public/ops/tree/main/juju/deployments) (if they are updated....).

Big if, being my point ;)

Doesn't stop this PR though, we should just make sure someone checks all deployments before the next revision starts going out.

File: src/service_args.py [hunk 3]
```diff
@@ -114,44 +114,10 @@ def __customize_service_args(self):
...
-        elif self.chain_name == 'turing':
-            self.__turing()
-        elif self.chain_name == 'bajun':
-            self.__bajun()
-        elif self.chain_name == 'joystream':
-            self.__joystream()
-        elif self.chain_name == 'equilibrium':
-            self.__equilibrium()
         elif self.chain_name.startswith('aleph-zero'):
             self.__aleph_zero()
```
Comment (jakobilobi) [inline src/service_args.py:118]: Should this not be removed anymore? Also, same thing further down.

**Comment by ZeeshanM96** [comment]
@jakobilobi @jonathanudd @awnns, As per the discussion we had here, I have double-checked all the chains with config URLs we have in the related models. I only couldn't confirm for `bajun` and `joystrream` because I couldn't find related models in our controllers. Also, just to be on the safe side, I am going to add chain spec urls to the collective pages of the chains as mentioned by @jonathanudd

**Comment by ZeeshanM96** [comment]
> > As per the discussion we had here, I have double-checked all the chains with config URLs we have in the related models. I only couldn't confirm for bajun and joystrream because I couldn't find related models in our controllers.
> 
> Awsome 👍 I'm pretty sure we don't run Bajun and Joytstream so no issue.
> 
> > Also, just to be on the safe side, I am going to add chain spec urls to the collective pages of the chains as mentioned by @jonathanudd. Url's can be found [here](https://nextcloud.dwellir.com/apps/collectives/Chain%20operations/Polkadot%20Ecosystem/Chain%20Spec%20URL's%20for%20Polkadot?fileId=92171).
> 
> There are subpages for each chain. Could you please add these urls in the sub pages instead? Feel free to ask if there are any chains you don't find.

Thanks @jonathanudd , I have also added the chain spec URLs in the sub-pages as well. I couldn't find chain names for `Bajun` and `Joystream`

**Comment by Maharacha** [comment]
> I think we can address the ones just replacing the name in an other card and/or PR. I'm not sure why it's designed like that.

It's because the chain_name is used for pulling a docker image right? But since a chain spec is not used, the `--chain` arg needs to be changed before starting the node. I mean e.g. only setting `mainnet` would not know which docker image to pull.

Should we remove the docker images from the charm as well? And set the whole image name in the config instead of only the tag?

File: src/service_args.py [hunk 4]
```diff
@@ -73,15 +73,14 @@ def __check_service_args(self, service_args: str or list):
...
-
         if msg:
             raise ValueError(msg)
 
     def __service_args_to_list(self, service_args: str) -> list:
         # Split on any number of spaces and '='. Hence, this will support both '--key value' and '--key=value' in the config.
         service_args = re.split(' +|=', service_args)
         return service_args
-
+
```
Comment (Maharacha) [inline src/service_args.py:83]: trailing whitespaces

File: src/service_args.py [hunk 5]
```diff
@@ -162,131 +132,32 @@ def __customize_service_args(self):
...
         # Update the URL to the master branch when the spec file is merged.
         chain_json_url = 'https://raw.githubusercontent.com/ajuna-network/Ajuna/el/tidy-chain-specs/resources/bajun/bajun-raw.json'
         chain_json_path = f"{c.CHAIN_SPEC_PATH}/bajun-raw.json"
-
         if not exists(chain_json_path):
             utils.download_chain_spec(chain_json_url, 'bajun-raw.json')
-
         self.__set_chain_name(chain_json_path, 0)
-
+
```
Comment (Maharacha) [inline src/service_args.py:144]: trailing whitespaces

**Comment by awnns** [comment]
> > I think we can address the ones just replacing the name in an other card and/or PR. I'm not sure why it's designed like that.
> 
> It's because the chain_name is used for pulling a docker image right? But since a chain spec is not used, the `--chain` arg needs to be changed before starting the node. I mean e.g. only setting `mainnet` would not know which docker image to pull.
> 
> Should we remove the docker images from the charm as well? And set the whole image name in the config instead of only the tag? What is weird with that is that we would still need image-specific logic in the charm code for what to extract from the image... But we could parse the image name instead for that logic. And then have something smart that could work for generic cases if a image logic is not implemented, or simply fail saying that image is not supported by the charm.

I think this would be inadvisable since it will induce a lot of extra work with the 150 (?) or so models we have already deployed. We could introduce an override as a temporary measure if it is necessary.

**Comment by Maharacha** [comment]
> > > I think we can address the ones just replacing the name in an other card and/or PR. I'm not sure why it's designed like that.
> > 
> > 
> > It's because the chain_name is used for pulling a docker image right? But since a chain spec is not used, the `--chain` arg needs to be changed before starting the node. I mean e.g. only setting `mainnet` would not know which docker image to pull.
> > Should we remove the docker images from the charm as well? And set the whole image name in the config instead of only the tag? What is weird with that is that we would still need image-specific logic in the charm code for what to extract from the image... But we could parse the image name instead for that logic. And then have something smart that could work for generic cases if a image logic is not implemented, or simply fail saying that image is not supported by the charm.
> 
> I think this would be inadvisable since it will induce a lot of extra work with the 150 (?) or so models we have already deployed. We could introduce an override as a temporary measure if it is necessary.

It would be a deprecation for while yes. Same as we have done for this chain spec transition. But this is of course out of scope for this PR.

**Comment by ZeeshanM96** [comment]
@jakobilobi @jonathanudd  I have removed `bajun` and `joystream` from the charm.

**Comment by jakobilobi** [comment]
> rebase or merge from main

From `develop` though, right?

**Comment by ZeeshanM96** [comment]
> > > rebase or merge from main
> > 
> > 
> > From `develop` though, right?
> 
> Yes sorry @ZeeshanM96 Should be from develop.

it's done @jakobilobi  and @jonathanudd

File: src/service_args.py [hunk 6]
```diff
@@ -161,132 +129,18 @@ def __customize_service_args(self):
             utils.download_chain_spec(self._local_relaychain_spec_url, 'relaychain-spec.json')
             self.__set_chain_name(f'{c.CHAIN_SPEC_DIR}/relaychain-spec.json', 1)
         if self._runtime_wasm_override:
-            self.__add_firstchain_args(['--wasm-runtime-overrides', c.WASM_DIR])
```
[the remark this answers is not in the document]
Comment (ZeeshanM96) [reply src/service_args.py:164]: @jonathanudd sorry i guess i skipped it during the merge, Please have a look now

**Pull request merged**